prov/efa/src/efa_mr.c

```diff
@@ -239,12 +239,18 @@ static int efa_mr_hmem_setup(struct efa_mr *efa_mr,
 				return err;
 			}
 		}
-		err = cuda_dev_register((struct fi_mr_attr *)attr, &efa_mr->peer.device.cuda);
-		if (err) {
-			EFA_WARN(FI_LOG_MR,
-				 "Unable to register handle for GPU memory. err: %d buf: %p len: %zu\n",
-				 err, attr->mr_iov->iov_base, attr->mr_iov->iov_len);
-			return err;
+
+		/* When cudaMemcpy is allowed (cuda_xfer_setting==CUDA_XFER_ENABLED), use cudaMemcpy (which does not need registration)
+		 * Otherwise, use gdrcopy and register the memory with gdrcopy.
+		 */
+		if (cuda_get_xfer_setting()!=CUDA_XFER_ENABLED) {
```

Comment:
Although registration for cudamemcpy itself is not required, I think we may depend on the registration time to allow libfabric to set CU_POINTER_ATTRIBUTE_SYNC_MEMOPS for the allocation so that we can ensure our memcpy's are synchronous.

Reply:
Not sure I understand.

The call to sync_memops happens inside efa provider:
https://github.com/ofiwg/libfabric/blob/1a25af05d48ed49cf3c3075c0df99f4a09b0068c/prov/efa/src/efa_mr.c#L235

Reply:
ok yes, there it is 👍